src/obstacle_detector.cpp

```diff
@@ -158,14 +165,70 @@ void ObstacleDetector::sensor_callback(const sensor_msgs::Range::ConstPtr &msg)
     else {
         RangeSensor& sensor = sensors[frame];
         sensor.update(msg->range, msg->header.stamp);
-        draw_line(sensor.origin, sensor.left_vertex, 0, 0, 1, sensor.id + 200);
-        draw_line(sensor.origin, sensor.right_vertex, 0, 1, 0, sensor.id + 300);
-        draw_line(sensor.left_vertex, sensor.right_vertex, 0.5, 0.5, 0,
+        draw_line(sensor.origin, sensor.left_vertex, 0.5, 0.5, 0.5, sensor.id + 200);
+        draw_line(sensor.origin, sensor.right_vertex, 0.5, 0.5, 0.5, sensor.id + 300);
+        draw_line(sensor.left_vertex, sensor.right_vertex, 1, 1, 1,
             sensor.id + 400);
     }
     obstacle_mutex.unlock();
 }
 
+void ObstacleDetector::scan_callback(const sensor_msgs::LaserScan::ConstPtr &msg)
+{
+    float theta = msg->angle_min;
+    float increment = msg->angle_increment;
+    float range_min = msg->range_min;
+
+    if (!have_lidar) {
+        try {
+            geometry_msgs::TransformStamped tfs =
+                tf_buffer->lookupTransform("base_link", "laser", ros::Time(0));
+
+            tf2::Transform tf;
+
+            // lidar origin
+            geometry_msgs::PointStamped origin;
+            origin.point.x = 0;
+            origin.point.y = 0;
+            origin.point.z = 0;
+            geometry_msgs::PointStamped base_origin;
+            tf2::doTransform(origin, base_origin, tfs);
+            fromMsg(base_origin.point, lidar_origin);
+
+            // normal vector
+            geometry_msgs::Vector3Stamped normal;
+            normal.vector.x = 1.0;
+            normal.vector.y = 0.0;
+            normal.vector.z = 0.0;
+            geometry_msgs::Vector3Stamped base_normal;
+            tf2::doTransform(normal, base_normal, tfs);
+            fromMsg(base_normal.vector, lidar_normal);
```

Comment:
More of a question:
Why do we need the normal?

I see it is being used in the Polar to xy computation, but I don't understand why

Reply:
It allows us to have the lidar in any orientation, instead of assuming that it is pointing forward, as the previous code did.